java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ScanOptions.java

```diff
@@ -26,12 +26,16 @@
 
   /**
    * Constructor.
-   * @param columns Projected columns. Empty for scanning all columns.
    * @param batchSize Maximum row number of each returned {@link org.apache.arrow.vector.ipc.message.ArrowRecordBatch}
+   * @param columns Projected columns. Null for scanning all columns.
    */
-  public ScanOptions(String[] columns, long batchSize) {
-    this.columns = columns;
+  public ScanOptions(long batchSize, String[] columns) {
```

Comment:
Why do you change the API?

Reply:
`columns` becomes an optional parameter after this PR. I tended to put it at tailing position of the parameter list. Regarding the API I would like to explain to users that they are all all experimental in [another PR](https://github.com/apache/arrow/pull/10333/files#diff-d83ff4a862278ef11112853a74fc9c0feb2f52ab17c8ffce81c1c84d2a50e5b1R24). But would you suggest that we should not make breaking changes at this time?

Reply:
I see. I was curious about the compatibility with the existing program.   
But, changing the major version number could break API compatibility.

Reply:
instead of allowing for null would it make sense to use Optional?

Reply:
Just to clarify, would you suggest to change the constructor to something like:

```Java
public ScanOptions(long batchSize, Optional<String[]> columns)
```

or 
```Java
public ScanOptions(Optional<String[]> columns, long batchSize)
```
? The issue we are trying to solve here is to make empty `columns` select zero columns rather than all columns (previous behavior). I don't think we can do anything to keep the compatibility when empty `columns` is specified.

Or by backwards compatibility you meant the compatibility in future changes? Then I don't have strong preference on using Optionals or nulls. Or maybe adding a builder pattern will also help? We will certainly have more options in future.

Reply:
By backwards compatibility I mean not breaking existing code written again the API.  In this case since it is easy being friendly to consumers makes sense.

I'm suggesting mark this constructor as deprecated but don't change the ordering of the fields.  Update the implementation here using one of the  new constructors proposed (I prefer the first option). IIUC the update would be passing optional.absent to the new constructor when columns.length == 0. 

This approach allows anybody using the existing constructor to still function and new users can use the constructor that takes the optional parameter with the new semantics.

I'd wait on the builder pattern until we have to add more parameters but ultimately that seems like a good idea.

Reply:
Thanks for clarification. I've made changes accordingly based on my understanding. Would you like to take another look?

Reply:
I leaved the shortcut constructor there. Should we remove it?

https://github.com/apache/arrow/pull/10652/commits/04378a103a2c1e9dbd822d7a18a7e8d651fff898#diff-522d7e8795ed981742735d48504889e81557c7a2f0e3a5a06368d2ca48fce36cR61-R63